### PR TITLE
Track already visited pairs of Iterables to avoid stack overflow

### PR DESCRIPTION
**Pull request by thomaszub**
Issue: #2915

## Overview

The visited `Iterable` are tracked and the assert fails if a already visited `Iterable` is visited again. This avoids a StackOverflowError.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java [hunk 1]
```diff
@@ -69,24 +79,39 @@ private static void assertIterableEquals(Iterable<?> expected, Iterable<?> actua
...
+	private static void assertIterablesNotAlreadyVisited(Iterable<?> expected, Iterable<?> actual,
+			Deque<Integer> indexes, Object messageOrSupplier, Collection<Iterable<?>> visitedIterablesExpected,
+			Collection<Iterable<?>> visitedIterablesActual) {
+		if (visitedIterablesExpected.contains(expected) || visitedIterablesActual.contains(actual)) {
+			failIterablesNotEqual(expected, actual, indexes, messageOrSupplier);
+		}
+		else {
+			visitedIterablesExpected.add(expected);
+			visitedIterablesActual.add(actual);
+		}
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java:112]: Can't this be simplified as follows?

```suggestion
		if (!visitedIterablesExpected.add(expected) && !visitedIterablesActual.add(actual)) {
			failIterablesNotEqual(expected, actual, indexes, messageOrSupplier);
		}
```

**Review by sbrannen — CHANGES_REQUESTED** [review]

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java (hunk 1 above)
Comment (thomaszub) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java:112]: Good point, will change it (the check must be "or": `!visitedIterablesExpected.add(expected) || !visitedIterablesActual.add(actual)`

Comment (thomaszub) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java]: 😁

Comment (thomaszub) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertIterableEquals.java]: Hi @sbrannen, could you please mark the request as solved. I force pushed the changes (maybe not best idea from my side). Thanks!

**Review by sbrannen — APPROVED** [review]

**Comment by marcphilipp** [comment]
I added a test with a counterexample to the proposed tracking of input values in e196f8d and added a new implementation that keeps track of pairs of iterables that are under investigation and potentially reuses the result should the same iterables be encountered again.

**Review by marcphilipp — APPROVED** [review]

**Review by sbrannen — CHANGES_REQUESTED** [review]
I think this looks good, but it would beneficial to add some inline documentation for how the algorithm currently works with regard to the "investigated elements" tracking. 

Also, the `SAME_ELEMENTS` enum constant could be renamed to something like `CONTAIN_SAME_ELEMENTS` or similar.

**Comment by marcphilipp** [comment]
> I think this looks good, but it would beneficial to add some inline documentation for how the algorithm currently works with regard to the "investigated elements" tracking.

Done in 6cf3516ff209c980aa63e65960847e820b939c0c.



> Also, the `SAME_ELEMENTS` enum constant could be renamed to something like `CONTAIN_SAME_ELEMENTS` or similar.

Done in 0c175abde477b92ba8ee69bca0b0b2e4a8093155.

**Comment by marcphilipp** [comment]
@thomaszub Thanks for the PR! 👍

**Pull request merged**